apps/prairielearn/src/pages/studentInstanceQuestionExam/studentInstanceQuestionExam.ejs

```diff
@@ -72,7 +73,5 @@
         </div>
       </div>
     </main>
-    <!-- Needs to be located below element scripts -->
```

Comment:
Do you know why this was? Is this not true/relevant anymore?

Reply:
I found this: https://github.com/PrairieLearn/PrairieLearn/pull/2367#issuecomment-643531459. I'll test the new code with this in mind.

Reply:
Ok, the issue here is that saving the form data is done in the ready event, but there are other initializations (particularly for elements) that also happen for the same event, and these initializations may update the inputs. The form data being saved should take these initializations into account. For example, on the initialization of pl-file-editor, the editor updates its hidden input with the base64 representation of the content.

In order to ensure that the question form data is saved after the initialization of input by the elements, the saving trigger must happen after the element initialization. The way the code used to take care of this was to add the code above to the bottom of the page, which would indeed cause this ready callback to be the last to be added to the event handler queue, and assuming the event handlers run in order (which I assume is a good assumption) this would ensure the saving happens at the end of the initialization.

In this case, since the call to `confirmOnUnload` is no longer guaranteed to be the last handler, I've changed the code to delay the saving of the form data with a [`setTimeout` with a delay of 0](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Event_loop#zero_delays). This causes the code to be executed in the next event cycle, i.e., after all other handlers that are already registered complete. I tested this in Chrome, Edge and Firefox, and it works as expected.

If you want to test that this works as expected locally: open a question that includes some element that has input initialization on document ready (e.g., pl-drawing or pl-file-editor; I used element/fileEditor and demo/drawing/simpleTutorial). If you change any of the inputs and refresh, it should show a warning that there are unsaved changes (wording depends on browser). If however you interact with the page but don't change the elements, the refresh should no longer show a warning.